### DISCUSSION TITLE
Number vs Float/Int

### DISCUSSION THREAD

**Discussion by Namek** [mint-lang/mint#175] [Language Development]
AFAIK the language supports only the Number type for numbers. However, quite often I define indices in my code. I want to make sure that index is an integer. Another case is having an amount of items.

Are you against having this distinction and if yes then why?

**Comment by gdotdesign** [comment]
The reason it only supports `Number` type because in JavaScript there is only [number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number). 

I think `Number` is good in the general but I'm not against having an `Int` (or `Integer`) type beside that. Implementation wise I don't know if it's enough to have a module for it or it needs compiler support.

**Comment by Namek** [comment]
Initially I thought about splitting between `Float` and `Int` but having the current `Number` and an additional `Int` sounds good to me.

I propose:
1. support `Int` as a type, I don't think that a sole module is enough
1. redefine `Math.floor()`/`Math.ceil()` to return `Int`
1. when doing math (`+`, `-`, `/`, `*`, `%`) between `Number` and `Int` would require an explicit conversion. `Int.toNumber(n : Number) : Int` and the Math functions from above. Maybe additional `Number.toInt(n : Number) : Int`.

**Comment by gdotdesign** [comment]
I think what you suggest can be done relatively easily, what we have to think about what operations would return in each case, this is what I got so far:
- `+`, `-`, `*` if any operands are `Number` it would return `Number`, if all operants are `Int` it would return `Int`
- `/` would only work on `Number / Number`
- `//` would only work on `Int / Int` (Integer division. The remainder is discarded.) (needs to be a function in the runtime)
- `%` I need to look it up how it behaves in other languages

In addition it would be nice to have an operator for https://en.wikipedia.org/wiki/Modular_arithmetic

`NumberLiterals` would have the `Number` type unless otherwise stated `property index : Int = 0` (in this case if a decimal is specified it would raise an error)

Under the hood it would all use the JavaScript `Number` type unless we use [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) which I don't recommend since it's support is only 70% https://caniuse.com/#feat=bigint

**Comment by Sija** [comment]
If we're to introduce `Int` type, I'd strongly suggest to add a `Float` counterpart - and both would descend from `Number`.

Distinction `Int` / `Number` on its own seems counter-intuitive.

**Comment by Namek** [comment]
> If we're to introduce `Int` type, I'd strongly suggest to add a `Float` counterpart - and both would descend from `Number`.
> 
> Distinction `Int` / `Number` on its own seems counter-intuitive.


@Sija Can you explain why would that be counter-intuitive?

**Comment by Sija** [comment]
> @Sija Can you explain why would that be counter-intuitive?

Because they mix the entities from two different levels - `Number` can be seen as the most general type, whereas `Int`, `Float` and possibly others, like `BigDecimal` are more concrete variants of it. Does saying out loud `Number.toInt` makes sense? That's what I meant by _counter-intuitive_. `Float.toInt` OTOH sounds pretty obvious to me re: the logic and function behind.

**Reply by danhowardms** [reply]
I disagree with this.  I think adding `Float` as a separate thing would add more confusion because it would be identical to `Number`.  You have a perfect one-to-one correspondence between all the possible Floats and all the possible Numbers, and the exact same behaviour in functions and operations.

`Int` and `Float` are not at the same 'level' as each other, because every `Int` is also a `Float` but only some `Float` values are also `Int`.

`Float` and `Number` *are* at the same 'level' because every `Float` is a `Number` and every `Number` is also a `Float`.

The distinction between `Int` and `Float` in other languages reflects the underlying binary representation, but that distinction is invalid here where the underlying representation for both would be javascript's `Number` (which is an IEEE754 double-precision float).

> Does saying out loud `Number.toInt` makes sense? That's what I meant by counter-intuitive. `Float.toInt` OTOH sounds pretty obvious to me

Neither of those sound very logical to me.  What would `Float.toInt(0.4)` be?  `floor()`, `ceil()` and `round()` make perfect sense when applied to any `Number` and always give you a `Number` which happens to also be an `Int`.
